File: packages/observability/src/ChdbClient.ts
/**
 * chDB (embedded ClickHouse) driver for Effect SQL, backed by chdb-node's
 * synchronous `Session`. Structure mirrors `@effect/sql`'s ClickhouseClient so
 * the `sql` tag, SqlSchema, Model and migrations work unchanged.
 *
 * Two encodings, kept separate:
 *   - WIRE VALUE (JSONEachRow): a real, round-trippable Schema codec. Results
 *     decode through it; insert rows encode through it. No SQL literals.
 *   - SQL LITERAL (query params only, since chDB has no bind channel): one-way
 *     code-gen, NOT a Schema (it can't round-trip). Composition is function
 *     composition via the `Ch` combinators below — no regex over type strings.
 *
 * Other chDB realities: in-process + SYNCHRONOUS (Effect.try, not tryPromise);
 * no pool / AbortController / KILL QUERY / server query_id; results are format
 * strings we parse; errors are thrown, with the ClickHouse code recovered from
 * the "Code: NNN. DB::Exception" message.
 *
 * VERIFY against effect/unstable/sql: Connection method set, Client.make
 * options, and Statement.makeCompiler/custom config are taken from the official
 * driver source; generics may need minor tweaks against the actual type defs.
 */
import {
  type ChdbQueryStream,
  type ChdbResult,
  type InsertParams,
  type InsertSummary,
  type QueryOptions,
  Session,
  type StreamOptions
} from "chdb"
import * as Context from "effect/Context"
import * as Effect from "effect/Effect"
import * as Layer from "effect/Layer"
import * as Schema from "effect/Schema"
import type * as Scope from "effect/Scope"
import * as Stream from "effect/Stream"
import * as Reactivity from "effect/unstable/reactivity/Reactivity"
import * as Client from "effect/unstable/sql/SqlClient"
import type { Connection, Row } from "effect/unstable/sql/SqlConnection"
import {
  ConnectionError,
  SqlError,
  SqlSyntaxError,
  StatementTimeoutError,
  UnknownError
} from "effect/unstable/sql/SqlError"
import * as Statement from "effect/unstable/sql/Statement"

const ATTR_DB_SYSTEM_NAME = "db.system.name"
const ATTR_DB_NAMESPACE = "db.namespace"

const escapeString = (s: string): string => s.replace(/\\/g, "\\\\").replace(/'/g, "\\'")

// ── error classification (chDB throws; code is in the message) ───────────────

const codeFromMessage = (cause: unknown): number | undefined => {
  const msg = cause instanceof Error ? cause.message : String(cause)
  const m = /Code:\s*(\d+)/.exec(msg)
  return m !== null ? Number(m[1]) : undefined
}

const SYNTAX_CODES = new Set([36, 47, 60, 62, 81, 242])
const classifyError = (
  cause: unknown,
  message: string,
  operation: string,
  fallback: "connection" | "unknown" = "unknown"
) => {
  const props = { cause, message, operation }
  const code = codeFromMessage(cause)
  if (code !== undefined) {
    if (SYNTAX_CODES.has(code)) return new SqlSyntaxError(props)
    if (code === 159 || code === 160 || code === 209) return new StatementTimeoutError(props)
  }
  return fallback === "connection" ? new ConnectionError(props) : new UnknownError(props)
}

// ── Ch: typed SQL-literal builders (for query params; one-way by design) ─────
// Each value knows its ClickHouse type and how to render a JS value as a
// literal. `array`/`nullable`/`map` compose by function composition, so
// `Ch.array(Ch.nullable(Ch.Int64))` builds both the renderer and the type with
// no regex and no stringly-typed dispatch.

export interface Ch<A> {
  readonly type: string
  readonly lit: (value: A) => string
}

const leaf = <A>(type: string, lit: (value: A) => string): Ch<A> => ({ type, lit })

export const Ch = {
  String: leaf<string>("String", (v) => `'${escapeString(v)}'`),
  Int64: leaf<number | bigint>("Int64", (v) => (typeof v === "bigint" ? v.toString() : String(Math.trunc(v)))),
  UInt64: leaf<number | bigint>("UInt64", (v) => (typeof v === "bigint" ? v.toString() : String(Math.trunc(v)))),
  Float64: leaf<number>("Float64", (v) => String(v)),
  Bool: leaf<boolean>("Bool", (v) => (v ? "true" : "false")),
  /** Pass epoch nanoseconds; rendered losslessly and timezone-independently. */
  DateTime64Nanos: leaf<number | bigint>(
    "DateTime64(9)",
    (v) => `fromUnixTimestamp64Nano(${typeof v === "bigint" ? v.toString() : String(v)})`
  ),
  UUID: leaf<string>("UUID", (v) => `'${escapeString(v)}'`),
  array<A>(inner: Ch<A>): Ch<ReadonlyArray<A>> {
    return leaf(`Array(${inner.type})`, (xs) => `[${xs.map(inner.lit).join(", ")}]`)
  },
  nullable<A>(inner: Ch<A>): Ch<A | null | undefined> {
    return leaf(`Nullable(${inner.type})`, (v) => (v === null || v === undefined ? "NULL" : inner.lit(v)))
  },
  map<K, V>(k: Ch<K>, v: Ch<V>): Ch<ReadonlyMap<K, V> | Record<string, V>> {
    return leaf(`Map(${k.type}, ${v.type})`, (m) => {
      const entries = m instanceof Map ? [...m.entries()] : Object.entries(m as Record<string, V>)
      return `map(${entries.flatMap(([mk, mv]) => [k.lit(mk as K), v.lit(mv as V)]).join(", ")})`
    })
  }
} as const

/** Default literal for untyped `sql` interpolation: dispatch on the JS runtime
 *  value (not on a parsed type string). For anything precise, use Ch + param. */
const defaultLiteral = (value: unknown): string => {
  if (value === null || value === undefined) return "NULL"
  if (Array.isArray(value)) return `[${value.map(defaultLiteral).join(", ")}]`
  switch (typeof value) {
    case "number":
      return (Number.isInteger(value) ? Ch.Int64 : Ch.Float64).lit(value)
    case "bigint":
      return Ch.Int64.lit(value)
    case "boolean":
      return Ch.Bool.lit(value)
    case "string":
      return Ch.String.lit(value)
    case "symbol":
      return Ch.String.lit(value.description ?? value.toString())
    case "function":
      return Ch.String.lit(value.name)
    case "undefined":
      return "NULL"
    case "object":
      if (value instanceof Date) return `fromUnixTimestamp64Milli(${value.getTime()})`
      if (value instanceof Map) {
        return `map(${
          Array.from(value.entries(), ([key, item]) => `${defaultLiteral(key)}, ${defaultLiteral(item)}`).join(", ")
        })`
      }
      return Ch.String.lit(JSON.stringify(value) ?? Object.prototype.toString.call(value))
  }
  return "NULL"
}

// ── compiler: the official ClickHouse one, but it emits final literals ───────
// (chDB can't bind params, so there is nothing to substitute later)

type ChdbLit = Statement.Custom<"ChdbLit", string, undefined> // paramA = pre-rendered literal
export type ChdbCustom = ChdbLit
const chdbLit = Statement.custom<ChdbLit>("ChdbLit")

const escape = Statement.defaultEscape("\"")

export const makeCompiler = (transform?: (_: string) => string) =>
  Statement.makeCompiler<ChdbCustom>({
    dialect: "sqlite",
    placeholder(_i, u) {
      return defaultLiteral(u) // render the literal in place; no {pN: Type}
    },
    onIdentifier: transform !== undefined
      ? (value, withoutTransform) => (withoutTransform ? escape(value) : escape(transform(value)))
      : escape,
    onRecordUpdate() {
      return ["", []]
    },
    onCustom(type) {
      return [type.paramA, []] // emit the pre-rendered literal, bind nothing
    }
  })

// ── fiber refs ────────────────────────────────────────────────────────────────

export const ClientMethod = Context.Reference<"query" | "command">("@chdb/ChdbClient/ClientMethod", {
  defaultValue: () => "query" as const
})

// ── config / model ───────────────────────────────────────────────────────────

export interface ChdbClientConfig {
  /** Session directory. A temp dir is created and removed on scope close if omitted. */
  readonly path?: string
  readonly database?: string
  /** Session-scoped settings, applied via SET (sticky for the Session's lifetime). */
  readonly settings?: Record<string, string | number | boolean>
  readonly spanAttributes?: Record<string, unknown>
  readonly transformResultNames?: (str: string) => string
  readonly transformQueryNames?: (str: string) => string
}

export class ChdbSession extends Context.Service<ChdbSession, Session>()(
  "@firegrid/observability/ChdbClient/ChdbSession"
) {}

export interface ChdbNative {
  readonly query: (sql: string, format?: string) => Effect.Effect<string, SqlError>
  readonly queryBind: (sql: string, args: object, format?: string) => Effect.Effect<string, SqlError>
  readonly queryAsync: (sql: string, options?: QueryOptions) => Effect.Effect<ChdbResult, SqlError>
  readonly queryBindAsync: (sql: string, params: object, options?: QueryOptions) => Effect.Effect<ChdbResult, SqlError>
  readonly insert: (params: InsertParams) => Effect.Effect<InsertSummary, SqlError>
  readonly queryStream: (sql: string, options?: StreamOptions) => Effect.Effect<ChdbQueryStream, SqlError>
}

export interface ChdbClientApi extends Client.SqlClient {
  readonly config: ChdbClientConfig
  readonly native: ChdbNative
  /** Typed literal fragment for a query parameter, e.g. param(Ch.Int64, 5n). */
  readonly param: <A>(ch: Ch<A>, value: A) => Statement.Fragment
  /** Run a statement and decode each row through a Schema (the typed-result path). */
  readonly query: <A, I>(
    schema: Schema.Codec<A, I>,
    statement: Effect.Effect<ReadonlyArray<unknown>, SqlError>
  ) => Effect.Effect<ReadonlyArray<A>, SqlError | Schema.SchemaError>
  /** INSERT row objects, encoded through a Schema to their JSONEachRow form. */
  readonly insertQuery: <A, I>(options: {
    readonly table: string
    readonly values: ReadonlyArray<A>
    readonly schema?: Schema.Codec<A, I>
    readonly format?: string
  }) => Effect.Effect<{ readonly written: number }, SqlError | Schema.SchemaError>
  readonly asCommand: <A, E, R>(effect: Effect.Effect<A, E, R>) => Effect.Effect<A, E, R>
}

export class ChdbClient extends Context.Service<ChdbClient, ChdbClientApi>()(
  "@firegrid/observability/ChdbClient"
) {}

const renderSetting = (v: string | number | boolean): string =>
  typeof v === "string" ? `'${escapeString(v)}'` : typeof v === "boolean" ? (v ? "1" : "0") : String(v)

// ── construction ─────────────────────────────────────────────────────────────

const openSession = (options: ChdbClientConfig): Effect.Effect<Session, SqlError, Scope.Scope> =>
  Effect.acquireRelease(
    Effect.try({
      // With no path, chdb-node opens its own temp directory and removes it on
      // cleanup(); with a path, the on-disk session persists across opens.
      try: () => (options.path === undefined ? new Session() : new Session(options.path)),
      catch: (cause) =>
        new SqlError({ reason: classifyError(cause, "ChdbClient: failed to open session", "connect", "connection") })
    }),
    // cleanup() is idempotent, never throws, and drops the temp dir when ephemeral.
    (s) => Effect.sync(() => s.cleanup())
  )

const bootstrapSession = (session: Session, options: ChdbClientConfig): Effect.Effect<void, SqlError> =>
  Effect.try({
    try: () => {
      if (options.database !== undefined) {
        session.query(`CREATE DATABASE IF NOT EXISTS ${options.database}; USE ${options.database}`)
      }
      if (options.settings !== undefined) {
        Object.entries(options.settings).forEach(([k, v]) => session.query(`SET ${k} = ${renderSetting(v)}`))
      }
      session.query("SELECT 1")
    },
    catch: (cause) =>
      new SqlError({ reason: classifyError(cause, "ChdbClient: bootstrap failed", "connect", "connection") })
  })

export const makeWithSession = (
  session: Session,
  options: ChdbClientConfig
): Effect.Effect<ChdbClientApi, SqlError, Reactivity.Reactivity> =>
  Effect.gen(function*() {
    const compiler = makeCompiler(options.transformQueryNames)
    const transformRows = options.transformResultNames !== undefined
      ? Statement.defaultTransforms(options.transformResultNames).array
      : undefined

    const runString = (sql: string, format: string): Effect.Effect<string, SqlError> =>
      Effect.try({
        try: () => String(session.query(sql, format) ?? ""),
        catch: (cause) => new SqlError({ reason: classifyError(cause, "Failed to execute statement", "execute") })
      })

    const native: ChdbNative = {
      query(sql, format = "CSV") {
        return Effect.try({
          try: () => String(session.query(sql, format) ?? ""),
          catch: (cause) => new SqlError({ reason: classifyError(cause, "Failed to execute native query", "execute") })
        })
      },
      queryBind(sql, args, format = "CSV") {
        return Effect.try({
          try: () => String(session.queryBind(sql, args, format) ?? ""),
          catch: (cause) =>
            new SqlError({ reason: classifyError(cause, "Failed to execute native bound query", "execute") })
        })
      },
      queryAsync(sql, queryOptions) {
        return Effect.tryPromise({
          try: () => session.queryAsync(sql, queryOptions),
          catch: (cause) =>
            new SqlError({ reason: classifyError(cause, "Failed to execute native async query", "execute") })
        })
      },
      queryBindAsync(sql, params, queryOptions) {
        return Effect.tryPromise({
          try: () => session.queryBindAsync(sql, params, queryOptions),
          catch: (cause) =>
            new SqlError({ reason: classifyError(cause, "Failed to execute native async bound query", "execute") })
        })
      },
      insert(params) {
        return Effect.tryPromise({
          try: () => session.insert(params),
          catch: (cause) => new SqlError({ reason: classifyError(cause, "Failed to execute native insert", "insert") })
        })
      },
      queryStream(sql, streamOptions) {
        return Effect.try({
          try: () => session.queryStream(sql, streamOptions),
          catch: (cause) =>
            new SqlError({ reason: classifyError(cause, "Failed to open native query stream", "execute") })
        })
      }
    }

    type TransformRows = NonNullable<Parameters<Connection["execute"]>[2]>

    const parseJson = (s: string): unknown => JSON.parse(s) as unknown
    const parseEachRow = (s: string): ReadonlyArray<Row> =>
      s.trim() !== ""
        ? s.trim().split("\n").map((line) => parseJson(line) as Row)
        : []
    const parseCompactEachRow = (s: string): ReadonlyArray<ReadonlyArray<unknown>> =>
      s.trim() !== ""
        ? s.trim().split("\n").map((line) => {
          const row = parseJson(line)
          return Array.isArray(row) ? row as ReadonlyArray<unknown> : [row]
        })
        : []
    const applyTransformRows = (
      rows: ReadonlyArray<Row>,
      transformRows: TransformRows | undefined
    ): ReadonlyArray<Row> => transformRows !== undefined ? transformRows(rows) : rows

    class ConnectionImpl implements Connection {
      // params are already inlined as literals by the compiler, so they're unused here
      private run(sql: string, _params: ReadonlyArray<unknown>, format = "JSONEachRow") {
        return Effect.withFiber<ReadonlyArray<Row>, SqlError>((fiber) => {
          if (fiber.getRef(ClientMethod) === "command") {
            return Effect.as(runString(sql, "Null"), [])
          }
          return Effect.map(runString(sql, format), parseEachRow)
        })
      }
      execute(sql: string, params: ReadonlyArray<unknown>, transformRows: TransformRows | undefined) {
        return Effect.map(this.run(sql, params), (rows) => applyTransformRows(rows, transformRows))
      }
      executeRaw(sql: string, _params: ReadonlyArray<unknown>) {
        return Effect.map(runString(sql, "JSON"), (s) => (s.trim() !== "" ? parseJson(s) : { data: [] }))
      }
      executeValues(sql: string, _params: ReadonlyArray<unknown>) {
        return Effect.map(runString(sql, "JSONCompactEachRow"), parseCompactEachRow)
      }
      executeValuesUnprepared(sql: string, params: ReadonlyArray<unknown>) {
        return this.executeValues(sql, params)
      }
      executeUnprepared(sql: string, params: ReadonlyArray<unknown>, transformRows: TransformRows | undefined) {
        return this.execute(sql, params, transformRows)
      }
      executeStream(sql: string, params: ReadonlyArray<unknown>, transformRows: TransformRows | undefined) {
        // chdb-node's Session is buffered, so this materializes then chunks.
        return this.run(sql, params).pipe(
          Effect.map((rows) => applyTransformRows(rows, transformRows)),
          Effect.map(Stream.fromIterable),
          Stream.unwrap
        )
      }
    }

    const connection = new ConnectionImpl()

    return Object.assign(
      yield* Client.make({
        acquirer: Effect.succeed(connection),
        compiler,
        spanAttributes: [
          ...(options.spanAttributes !== undefined ? Object.entries(options.spanAttributes) : []),
          [ATTR_DB_SYSTEM_NAME, "clickhouse"],
          [ATTR_DB_NAMESPACE, options.database ?? "default"]
        ],
        transformRows
      }),
      {
        config: options,
        native,
        param<A>(ch: Ch<A>, value: A) {
          return Statement.fragment([chdbLit(ch.lit(value), undefined)])
        },
        query<A, I>(schema: Schema.Codec<A, I>, statement: Effect.Effect<ReadonlyArray<unknown>, SqlError>) {
          const decode = Schema.decodeUnknownEffect(Schema.Array(schema))
          return Effect.flatMap(statement, (rows) => decode(rows))
        },
        insertQuery<A, I>(
          opts: {
            readonly table: string
            readonly values: ReadonlyArray<A>
            readonly schema?: Schema.Codec<A, I>
            readonly format?: string
          }
        ) {
          return Effect.gen(function*() {
            const rows: ReadonlyArray<unknown> = opts.schema !== undefined
              ? yield* Schema.encodeUnknownEffect(Schema.Array(opts.schema))(opts.values)
              : opts.values
            return yield* Effect.try({
              try: () => {
                const format = opts.format ?? "JSONEachRow"
                const data = rows.map((v) => JSON.stringify(v)).join("\n")
                session.query(`INSERT INTO ${opts.table} FORMAT ${format}\n${data}`)
                return { written: opts.values.length }
              },
              catch: (cause) => new SqlError({ reason: classifyError(cause, "Failed to insert data", "insert") })
            })
          })
        },
        asCommand<A, E, R>(effect: Effect.Effect<A, E, R>) {
          return Effect.provideService(effect, ClientMethod, "command")
        }
      }
    ) as ChdbClientApi
  })

export const make = (
  options: ChdbClientConfig
): Effect.Effect<ChdbClientApi, SqlError, Scope.Scope | Reactivity.Reactivity> =>
  Effect.gen(function*() {
    const session = yield* openSession(options)
    yield* bootstrapSession(session, options)
    return yield* makeWithSession(session, options)
  })

// ── layers ───────────────────────────────────────────────────────────────────

export const sessionLayer = (config: ChdbClientConfig): Layer.Layer<ChdbSession, SqlError> =>
  Layer.effectContext(
    Effect.gen(function*() {
      const session = yield* openSession(config)
      yield* bootstrapSession(session, config)
      return Context.make(ChdbSession, session)
    })
  )

export const layerFromSession = (
  config: ChdbClientConfig
): Layer.Layer<ChdbClient | Client.SqlClient, SqlError, ChdbSession> =>
  Layer.effectContext(
    Effect.gen(function*() {
      const session = yield* ChdbSession
      const client = yield* makeWithSession(session, config)
      return Context.make(ChdbClient, client).pipe(Context.add(Client.SqlClient, client))
    })
  ).pipe(Layer.provide(Reactivity.layer))

export const layer = (config: ChdbClientConfig): Layer.Layer<ChdbSession | ChdbClient | Client.SqlClient, SqlError> =>
  layerFromSession(config).pipe(Layer.provideMerge(sessionLayer(config)))

File: src/Firegrid.Trace/ChdbClient.fs
namespace Firegrid.Trace

open Effect
open Fable.Core
open Fable.Core.JsInterop

/// A pragmatic, REDUCED EffSharp chDB client.
///
/// `ChdbClient.ts` is built on `effect/unstable/sql` (SqlClient, the Statement
/// compiler, the Connection machinery) plus the `chdb` Node package. EffSharp has
/// no equivalent of effect/unstable/sql, so the Statement compiler / SqlClient
/// / Connection layer are intentionally NOT ported. What IS ported, faithfully:
///   - `ChdbClientConfig` and the session open/bootstrap lifecycle.
///   - The `ChdbNative` query surface (query / queryBind / queryAsync /
///     queryBindAsync / insert / queryStream) each returning an EffSharp
///     `Effect<_, SqlError, _>` and classifying thrown chDB errors via
///     `Ch.classifyError`.
///   - `param` / `insertQuery` helpers built on `Ch` and JSON.
/// `SqlError` is the reduced `Ch.SqlError` record (kind + cause + message +
/// operation), since effect/unstable/sql's `SqlError` is unavailable.
[<RequireQualifiedAccess>]
module ChdbClient =

    type SqlError = Ch.SqlError

    /// Session directory / database / sticky settings (faithful to the TS config,
    /// minus the SqlClient-only `transformResultNames`/`transformQueryNames` hooks
    /// which belonged to the omitted compiler layer; retained as inert options).
    type ChdbClientConfig =
        { /// Session directory. A temp dir is created and removed on close if omitted.
          Path: string option
          Database: string option
          /// Session-scoped settings applied via SET (sticky for the session).
          Settings: (string * obj) list
          SpanAttributes: (string * obj) list }

        static member Empty =
            { Path = None
              Database = None
              Settings = []
              SpanAttributes = [] }

    // ── try/classify bridge ──────────────────────────────────────────────────

    /// Run a synchronous chDB thunk; on throw classify the error and fail the
    /// Effect. Mirrors TS `Effect.try({ try, catch })`.
    let private trySync
        (thunk: unit -> 'a)
        (message: string)
        (operation: string)
        (fallback: Ch.SqlErrorKind)
        : Effect<'a, SqlError, 'R> =
        let outcome = TraceSdk.tryRun thunk

        if TraceSdk.prop<bool> outcome "ok" then
            Effect.succeed (TraceSdk.prop<'a> outcome "value")
        else
            let cause = TraceSdk.prop<obj> outcome "error"
            Effect.fail (Ch.classifyError cause message operation fallback)

    /// Wrap a chDB promise; on rejection classify and fail. Mirrors TS
    /// `Effect.tryPromise`.
    let private tryPromise
        (promise: unit -> JS.Promise<'a>)
        (message: string)
        (operation: string)
        : Effect<'a, SqlError, 'R> =
        Effect.tryPromiseJS promise (fun cause -> Ch.classifyError (box cause) message operation Ch.UnknownError)

    // ── settings rendering ───────────────────────────────────────────────────

    /// `string -> '...' ; boolean -> 1|0 ; number -> String(v)`.
    let private renderSetting (v: obj) : string =
        match TraceSdk.typeOf v with
        | "string" -> "'" + Ch.escapeString (unbox<string> v) + "'"
        | "boolean" -> if unbox<bool> v then "1" else "0"
        | _ -> TraceSdk.stringValue v

    // ── construction / lifecycle ─────────────────────────────────────────────

    /// `new Session()` / `new Session(path)` wrapped in an Effect. The TS uses
    /// `Effect.acquireRelease`; EffSharp here exposes the open + an explicit
    /// `closeSession` (call on scope close). cleanup() drops the temp dir.
    let openSession (options: ChdbClientConfig) : Effect<TraceSdk.RawSession, SqlError, 'R> =
        trySync
            (fun () ->
                match options.Path with
                | None -> TraceSdk.openSession ()
                | Some path -> TraceSdk.openSessionWithPath path)
            "ChdbClient: failed to open session"
            "connect"
            Ch.ConnectionError

    /// `session.cleanup()` — idempotent release.
    let closeSession (session: TraceSdk.RawSession) : Effect<unit, SqlError, 'R> =
        Effect.sync (fun () -> TraceSdk.cleanup session)

    /// CREATE DATABASE + USE (if set), apply session SET settings, then SELECT 1.
    let bootstrapSession (session: TraceSdk.RawSession) (options: ChdbClientConfig) : Effect<unit, SqlError, 'R> =
        trySync
            (fun () ->
                match options.Database with
                | Some db -> TraceSdk.query session ("CREATE DATABASE IF NOT EXISTS " + db + "; USE " + db) |> ignore
                | None -> ()

                for (k, v) in options.Settings do
                    TraceSdk.query session ("SET " + k + " = " + renderSetting v) |> ignore

                TraceSdk.query session "SELECT 1" |> ignore)
            "ChdbClient: bootstrap failed"
            "connect"
            Ch.ConnectionError

    // ── native query surface (the portable part of ChdbClientApi) ────────────

    /// Mirrors `ChdbNative`. Results are returned as the raw chDB result coerced
    /// to a string (sync paths) or the native result object (async/insert paths).
    type ChdbNative =
        { Query: string -> string -> Effect<string, SqlError, unit>
          QueryBind: string -> obj -> string -> Effect<string, SqlError, unit>
          QueryAsync: string -> obj -> Effect<obj, SqlError, unit>
          QueryBindAsync: string -> obj -> obj -> Effect<obj, SqlError, unit>
          Insert: obj -> Effect<obj, SqlError, unit>
          QueryStream: string -> obj -> Effect<obj, SqlError, unit> }

    /// `String(session.query(sql, format) ?? "")`.
    let private queryString (session: TraceSdk.RawSession) (sql: string) (format: string) : string =
        let raw = TraceSdk.queryFormat session sql format
        if TraceSdk.isNullish raw then "" else TraceSdk.stringValue raw

    let private queryBindString (session: TraceSdk.RawSession) (sql: string) (args: obj) (format: string) : string =
        let raw = TraceSdk.queryBind session sql args format
        if TraceSdk.isNullish raw then "" else TraceSdk.stringValue raw

    /// Build the native surface over an open session. (chDB query default format
    /// is "CSV", matching the TS defaults.)
    let makeNative (session: TraceSdk.RawSession) : ChdbNative =
        { Query =
            fun sql format ->
                trySync (fun () -> queryString session sql format) "Failed to execute native query" "execute" Ch.UnknownError
          QueryBind =
            fun sql args format ->
                trySync
                    (fun () -> queryBindString session sql args format)
                    "Failed to execute native bound query"
                    "execute"
                    Ch.UnknownError
          QueryAsync =
            fun sql options ->
                tryPromise
                    (fun () -> TraceSdk.queryAsync session sql options)
                    "Failed to execute native async query"
                    "execute"
          QueryBindAsync =
            fun sql parameters options ->
                tryPromise
                    (fun () -> TraceSdk.queryBindAsync session sql parameters options)
                    "Failed to execute native async bound query"
                    "execute"
          Insert =
            fun parameters ->
                tryPromise (fun () -> TraceSdk.insert session parameters) "Failed to execute native insert" "insert"
          QueryStream =
            fun sql options ->
                trySync
                    (fun () -> TraceSdk.queryStream session sql options)
                    "Failed to open native query stream"
                    "execute"
                    Ch.UnknownError }

    // ── reduced client API ───────────────────────────────────────────────────

    /// Options for `insertQuery`. Mirrors the TS `insertQuery` opts, minus the
    /// `schema` codec (Schema encoding belonged to effect's Schema, not ported).
    type ChdbInsertOptions =
        { Table: string
          Values: obj list
          Format: string option }

    /// The reduced replacement for `ChdbClientApi`. The SqlClient/`sql` tag, the
    /// Schema-decoding `query`, the compiler-backed `Statement.Fragment` plumbing
    /// and `asCommand` are dropped (they belonged to effect/unstable/sql). What
    /// remains is the native surface; literal/insert helpers are module functions
    /// (`param`, `insertQuery`).
    type ChdbClient =
        { Config: ChdbClientConfig
          Session: TraceSdk.RawSession
          Native: ChdbNative
          /// INSERT row objects as JSONEachRow (or `format`); returns rows written.
          InsertQuery: ChdbInsertOptions -> Effect<int, SqlError, unit> }

    /// Render a typed parameter literal: `param Ch.Int64 5L`. (The TS produced a
    /// `Statement.Fragment`; without the compiler we expose the raw SQL literal.)
    let param (ch: Ch<'A>) (value: 'A) : string = ch.Lit value

    /// `INSERT INTO table FORMAT <format>\n<ndjson>`; returns the count written.
    let insertQuery (session: TraceSdk.RawSession) (options: ChdbInsertOptions) : Effect<int, SqlError, unit> =
        trySync
            (fun () ->
                let format = options.Format |> Option.defaultValue "JSONEachRow"
                let data = options.Values |> List.map TraceSdk.stringify |> String.concat "\n"
                TraceSdk.query session ("INSERT INTO " + options.Table + " FORMAT " + format + "\n" + data) |> ignore
                List.length options.Values)
            "Failed to insert data"
            "insert"
            Ch.UnknownError

    /// Build a reduced `ChdbClient` over an existing session (mirrors
    /// `makeWithSession`, minus the SqlClient assembly).
    let makeWithSession (session: TraceSdk.RawSession) (options: ChdbClientConfig) : Effect<ChdbClient, SqlError, 'R> =
        Effect.succeed
            { Config = options
              Session = session
              Native = makeNative session
              InsertQuery = insertQuery session }

    /// Open + bootstrap + build the reduced client (mirrors `make`). The caller is
    /// responsible for `closeSession` on teardown.
    let make (options: ChdbClientConfig) : Effect<ChdbClient, SqlError, 'R> =
        effect {
            let! session = openSession options
            do! bootstrapSession session options
            return! makeWithSession session options
        }
